Docs build improvements

wadells opened the pull request:
### Summary

This PR introduces a new make target `docs-lint` which runs `milv -ignore-external` on all of our docs. As well, the regular `make docs` and `make run-docs` now use the `--strict` flag to catch issues.  Both of these changes are building towards enforcement of these targets on all pull requests.

Why?  We had tens to hundreds of broken **_internal_** links in our customer facing documentations that @benarent and I cleaned up in #1210. Providing tooling to catch link rot and enforcing that will prevent issues like this from creeping back in.  This also establishes an engineering norm that docs quality is enforced by automatic tools.

In addition to the necessary changes, I also refactored our docs Makefile substantially, to better use partial builds and caching.  See individual commit messages for more information about those changes.

Contributes to fixing #1220.

### Testing Done

`make docs` and `make docs-lint` on this branch before rebase:
[docs-build-dirty.txt](https://github.com/gravitational/gravity/files/4315484/docs-build-dirty.txt)

`make docs` and `make docs-lint` on this branch after rebase (updated):
[docs-build-clean.txt](https://github.com/gravitational/gravity/files/4320063/docs-build-clean.txt)

I also have a PR job basically ready to run these new targets against every PR to gravity.  See https://jenkins.gravitational.io/job/gravity-pr-docs/ and the test PRs here:

https://github.com/wadells/gravity/pull/1
https://github.com/wadells/gravity/pull/2
https://github.com/wadells/gravity/pull/3

We just need to flip it over from my personal repo to the main repo and configure webhook pushes to make it go live.

### Removal of sass

@alex-kovoy @benarent: I've removed sass from our docker file in 7528d552. This should be safe @alex-kovoy removed the last scss/sass files in 2e1724ed7e3c7742c7c32a114f028108f77bda7d:

```
walt@work:~/git/gravity/docs$ git log --all --oneline -- "*.sass"                                                                                                                             
walt@work:~/git/gravity/docs$ git log --all --oneline -- "*.scss"                                                                                                                             
2e1724ed (docs) cleanup                                                                                                                                                                       
b580aeac Wolfe/docs (#79)                                                                                                                                                                     
ff8241d9 (tag: 5.2.0-alpha.2) Move all OSS code to this repository.                                                                                                                           
45b74e98 (tag: 5.2.0-alpha.0) Add OSS code.                                                                                                                                                   
walt@work:~/git/gravity/docs$ git log --stat -1 2e1724ed                                                                                                                                      
                                                                                                                                                                                              
commit 2e1724ed7e3c7742c7c32a114f028108f77bda7d
Author: Alexey Kontsevoy <alex@kontsevoy.com>
Date:   Fri Nov 2 16:18:42 2018 -0400

    (docs) cleanup

 docs/build.sh                              |    10 +-
 docs/theme/css/highlight.css               |   125 --
 docs/theme/css/telekube-bundle.css         |    46 -
 docs/theme/css/theme.css                   |    12 -
 docs/theme/head.html                       |     3 +-
 docs/theme/js/base.js                      |    76 -
 docs/theme/js/grvlib.js                    |   619 --------
 docs/theme/js/highlight.pack.js            |     2 -
 docs/theme/js/jquery-2.1.1.min.js          |     4 -
 docs/theme/js/modernizr-2.8.3.min.js       |     1 -
 docs/theme/js/theme.js                     |   262 ----
 docs/theme/scripts.html                    |    16 +-
 docs/theme/{js/grvlib.css => src/docs.css} |    26 +
 docs/theme/src/docs.js                     | 11189 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 docs/theme/src/index.scss                  |    94 --
 docs/theme/{js => src}/prettify-yaml.js    |     0
 docs/theme/{js => src}/prettify.js         |     0
 docs/theme/{js => src}/third-party.js      |     0
 18 files changed, 11226 insertions(+), 1259 deletions(-)
```

wadells committed:
Refactor docs Makefile.

This commit brings in two primary changes:

 1. The mkdocs-buildbox is cached in a cookie, to avoid rebuilding

    This saves us time and a significant amount of shell transcript with
    each `make docs` or `make run`.

 2. `make run` now accepts `RUN_CFG=5.x.yaml` allowing the dev server to
    serve docs other than `latest.yaml`.  For example:

    `make run RUN_CFG=5.x.yaml`

wadells committed:
Remove vestigial Dockerfile directives.

All this logic was inherited when this image was initially copied from
teleport. With the exception of the teleport directories, both the LANG
values and DEBIAN_FRONTEND is handled in the upstream mkdocs-base:

  https://github.com/gravitational/ops/blob/2c096b225a52293891b142cb64c8ea33ae3c7a57/docker-images/mkdocs/Dockerfile#L14-L17

The teleport directories are unused for Gravity documentation builds.

wadells committed:
Add lint target to docs Makefile.

In order to help employees keep docs polished, add a docs lint target.

wadells committed:
Simplify repeated logic with DOCKERMAKE.

We run several make targets within docker.  Use a common executable
DOCKERMAKE to be sure these all get handled the same way.

wadells committed:
Build docs to gravity version build dir.

Instead of always building docs to ../build/docs, they are now built to
../build/$(VERSION)/docs when run from the top level makefile.  I'm a
bit worried this will confuse some tool that expects only binaries in
that directory.  Nonetheless, this allows for easier iteration on
different versions of the documentation.

As part of this, I needed to remove the $(BUILDDIR) target, because it
was constantly being updated by the build (rightfully so), and thus
re-triggering the $(DOCBOX_ID) target.

wadells committed:
Force --strict for building & developing docs.

No more warnings allowed. We're keeping our docs as clean as possible.

wadells committed:
Switch docker cache from --iidfile to touch.

Keep our docker build caching, but use touch to create the file instead
of --iidfile because jenkin's docker is a bit too old. See ops issue #141
for more information.

wadells commented on docs/Makefile:
```diff
@@ -1,41 +1,91 @@
 	docker build \
 		--build-arg UID=$$(id -u) \
 		--build-arg GID=$$(id -g) \
 		--build-arg HOME=$(HOME) \
```
Does this need the WORKDIR change from above?  It doesn't seem to matter to the end user.

wadells replied:
fixed in ee1376f

wadells commented on docs/Makefile:
```diff
@@ -1,41 +1,91 @@
+	cd $(*F) && milv -ignore-external
+
+.PHONY:$(LATEST_SYMLINK)
+$(LATEST_SYMLINK): $(LATEST_BUILD) $(BUILDDIR)
```
Neither of these dependencies makes sense.  I believe they're vestigial from some work discarded along the way.

a-palchikov replied:
Let's remove it then.

wadells replied:
removed in b4b64f6

wadells commented on docs/Dockerfile:
```diff
@@ -13,18 +13,15 @@ RUN groupadd jenkins --gid=$GID -o && useradd jenkins --uid=$UID --gid=$GID --cr
-    LANG="en_US.UTF-8" \
-    LC_ALL="en_US.UTF-8" \
-    LC_CTYPE="en_US.UTF-8" \
+RUN apt-get update -y && apt-get install -q -y ruby-sass parallel
```
parallel isn't needed anymore, with the death of build.sh

a-palchikov replied:
Same here - if it's not needed anymore - remove it.

wadells replied:
removed in b4b64f6.  I was able to go a bit further and remove sass (and thus all the apt directives) in 7528d55.

wadells commented on Makefile:
```diff
@@ -746,6 +746,11 @@ docs:
+docs-lint:
+	$(MAKE) -C docs lint
+
+
```
extra whitespace

a-palchikov replied:
I have to ask - is this a question or statement? Why not remove it?

wadells replied:
I will remove it.  I only saw this issue when I was putting the PR up at the end of the day.

wadells replied:
removed in b1a2ffd

wadells commented:
Macro question: Should the `docs-lint` target be separate from the `docs` target?  Or should `docs` depend on `docs-lint` to make this easier to use.

a-palchikov commented on docs/Dockerfile:
```diff
@@ -13,18 +13,15 @@ RUN groupadd jenkins --gid=$GID -o && useradd jenkins --uid=$UID --gid=$GID --cr
-    GOROOT="/opt/go" \
     PATH="$PATH:/opt/go/bin:/gopath/bin"
+RUN cd /opt && curl https://dl.google.com/go/go1.14.linux-amd64.tar.gz | tar xz
+RUN GO111MODULE=on go get -u -v github.com/magicmatatjahu/milv@v0.0.6
```
How about move all things only relevant for building the prerequisites (like Go for building milv here) to the build container (perhaps as a multistage build) and have the docs container without it?

wadells replied:
Good idea.  I'll look into this.

wadells replied:
Great suggestion.  The Dockerfile is considerably simpler after I implemented this change in 
9c8fd4a.

a-palchikov commented on docs/Makefile:
```diff
@@ -1,41 +1,91 @@
+# drops you into bash shell of mkdocs container
+.PHONY:shell
+shell: $(DOCBOX_ID_FILE)
+	docker run -ti $(NOROOT) $(DOCKERFLAGS) $(DOCKERPORTS) $(DOCBOX) \
```
Also `--rm`?

wadells replied:
`--rm=true` is in the definition of `DOCKERFLAGS` on line 11.

a-palchikov commented on docs/Makefile:
```diff
@@ -1,41 +1,91 @@
-# builds a docker container which is used for running `mkdocs`
-.PHONY:bbox
-bbox:
+.PHONY:docbox
```
What's `docbox`?

wadells replied:
Good catch.  I don't remember why this changed.  It should be returned to be consistent.

wadells replied:
removed in b4b64f6

wadells committed:
Refactor docs/Dockerfile to use a multi-step build.

As suggested by @a-palchkov here:

  https://github.com/gravitational/gravity/pull/1231/files/29a698efd88ccdf954ca4bfc5e2ef476590954a4#r390895629

This considerably simplifies our final docs image, keeping it smaller
and the logic for building it a fair bit cleaner.

wadells committed:
Remove sass from docbox.

@alex-kovoy removed the last .scss file from this repository in:

  2e1724ed7e3c7742c7c32a114f028108f77bda7d

There were never any .sass files. Thus we can remove sass and
dependencies from the docbox.

wadells committed:
Switch docbox user from jenkins -> builder.

This username is simply a placeholder for the actual user of the system.
It has nothing to do with jenkins. Having the name jenkins implies that
they're somehow related or interplay with each other... which is
misleading.

"builder" feels like "jenkins" (implying a user for build purposes) but
without the tie to a specific system.

I also took the opportunity to swap to bash instead of sh, as bash is
present, and provides a better UX.

wadells committed:
Swap HOME -> WORKDIR in docs/Makefile.

@benarent change the dockerfile to expect WORKDIR in #1210.  However he
didn't update the Makefile to accomodate this change. It doesn't matter
too much, since DOCKERFLAGS re-declares -w, -v, and -p anyhow. Still,
best to be consistent.
